ehsm_kms_service/function.js

```diff
@@ -506,6 +506,99 @@ const _checkParams = function (req, res, next, nonce_database, DB) {
   }
 }
 
+/**
+ * ehsm napi result
+ * If the value of the result is not equal to 200, the result is directly returned to the user
+ * @param {function name} action
+ * @param {NAPI_* function params} params
+ * @returns napi result | false
+ */
+function napi_result_local (action, params) {
+  try {
+    const napi_res = ehsm_napi[`NAPI_${action}`](...params)
+    let res = JSON.parse(napi_res)
+    if (res && res.code == 200) {
+      return res
+    }
+    return false
+  } catch (e) {
+    return false
+  }
+}
+
+/**
+ * Query ApiKey
+ */
+const _query_api_key = async (DB, appid) => {
+  try {
+    const db_query = {
+      selector: {
+        _id: `user_info:${appid}`,
+      },
+      fields: ['appid', 'apikey', 'cmk'],
+      limit: 1,
+    }
+    let query_result = await DB.partitionedFind('user_info', db_query);
+    if (!(query_result && query_result.docs[0])) {
+      return {
+        msg: 'Query User Info error',
+        api_key: ''
+      }
+    }
+    let { cmk, apikey } = query_result.docs[0]
+    let decypt_result = napi_result_local(
+      cryptographic_apis.Decrypt,
+      [cmk, apikey, '']
+    )
+    if (decypt_result) {
+      let decoded_api_key = base64_decode(decypt_result.result.plaintext_base64)
+      if (decoded_api_key) {
+        return {
+          msg: '',
+          api_key: decoded_api_key
+        }
+      } else {
+        return {
+          msg: 'Decode key error',
+          api_key: ''
+        }
+      }
+    } else {
+      return {
+        msg: 'Decrypt key error',
+        api_key: ''
+      }
+    }
+  } catch (error) {
+    logger.error(JSON.stringify(error))
+  }
+  return {
+    msg: 'Unexcept error',
+    api_key: ''
+  }
+}
+
+/**
+ * Gen Hmac 
+ */
+const gen_hmac = async (DB, appid, sign_params) => {

```

Comment:
Is it possible to use gen_hmac to replace the similar codes at: https://github.com/intel/ehsm/blob/main/ehsm_kms_service/function.js#L473